### ISSUE TITLE
About the requirement of using `Uint256` as `key` in `Map`.

### ISSUE THREAD

**Issue by 99Kies** [CosmWasm/cw-storage-plus#40]
I have a `Map` storage, and I need use `Uint256`-type to be my `key` type.

----

`state.rs`

```rust
pub const MSG_HASHES: Map<&Uint256, Uint256> = Map::new("msg_hashes");
```

`contract.rs`

```rust
    MSG_HASHES.save(
        deps.storage,
        &Uint256::from_u128(0u128),
        &Uint256::from_u128(0u128),
    )?;
```

![image](https://github.com/CosmWasm/cosmwasm/assets/45918704/1a38ede9-96d0-47d7-aa07-7c9dc58f1e37)

**Comment by webmaster128** [comment]
This is currently not possible. Maybe it could be implemented.

A workaround is to use `[u8; 32]` as the key and use Uint256::to_be_bytes/::from_be_bytes when accessing it.

**Comment by 99Kies** [comment]
If we want to add support for this, is Uint256 going to implement the PrimaryKey macro?

**Comment by 99Kies** [comment]
> This is currently not possible. Maybe it could be implemented.
> 
> A workaround is to use `[u8; 32]` as the key and use Uint256::to_be_bytes/::from_be_bytes when accessing it.

`[u8; 32]` have same error😂

![image](https://github.com/CosmWasm/cw-storage-plus/assets/45918704/3c50682a-f065-4568-b7ba-4156f497ee28)

**Comment by webmaster128** [comment]
Bummer

**Comment by 99Kies** [comment]
And I found that neither `Uint256` nor `HexString` are supported as keys, and that `Uint256` lacks import methods for `decimal` and `hex_string`.

**Comment by webmaster128** [comment]
You can search the codebase for `PrimaryKey`. There are a bunch of current implementations and explanations in a README.

I think we should start by implementing it for `[u8; N]` as this can serve as a basis for all kinds of other fixed-length types.

**Comment by maurolacy** [comment]
I would also think / reflect about why you need / want such a large integer type as key in the first place.

**Comment by webmaster128** [comment]
> i.e. using `(u128, u128)`.

But this introduces a separator in the underlying storage bytes, right? Having `[u8; N]` would avoid that.

**Comment by 99Kies** [comment]
> You should consider using a tuple for key as a wrapper / alternative to your `Uint256` type, i.e. using `(u128, u128)`.
> 
> Also, you can use raw `&[u8]` and `Vec<u8>` types as keys, AFAIK.

Why not consider adding the Uint256 type directly to the `PrimaryKey` macro

**Comment by 99Kies** [comment]
I've solved it this way for now.

```rust
pub const MSG_HASHES: Map<Vec<u8>, Uint256> = Map::new("msg_hashes");


MSG_HASHES.save(
    deps.storage,
    Uint256::from_u128(0u128).to_be_bytes().to_vec(),
    &Uint256::from_u128(0u128),
)?;
```

**Comment by 99Kies** [comment]
@maurolacy I think `Uint256` will be very helpful for developers moving from `solidity`, and there is actually quite a bit of content that needs uint256 to handle, such as `pubkey` data like this. (If uint256 could be used more elegantly, I believe it would be easier to migrate the business over on solidity.) 😁

**Comment by chipshort** [comment]
@99Kies `update` does not take the value directly. It wants a function that gets the current value as an input and returns the new one.
Please check the [example in the README](https://github.com/CosmWasm/cw-storage-plus/blob/3822f951ab33ad289e2b1333b6373a55b30906c7/README.md?plain=1#L245-L247)

**Comment by 99Kies** [comment]
@chipshort I solved it with this code.
```rust
NODES.update(
        deps.storage,
        parent_idx.to_be_bytes().to_vec(),
        |_c: Option<Uint256>| -> StdResult<_> { Ok(hash5(inputs)) },
    )
```

**Comment by chipshort** [comment]
Done in #44. Will be released as part of 1.1.

**Comment by webmaster128** [comment]
You can now use the type `[u8; 32]` as a Map key which makes your code much nicer since you avoid unwraps and can just convert with to_be_bytes and from_be_bytes.
